ext/linux/osq_lock.h

```diff
@@ -0,0 +1,322 @@
+
+    for (;;) {
+        if (atomic_read(&lock->tail) == curr &&
+            atomic_cmpxchg_acquire(&lock->tail, curr, old) == curr) {
+            /*
+             * We were the last queued, we moved @lock back. @prev
+             * will now observe @lock and will complete its
+             * unlock()/unqueue().
+             */
+            break;
+        }
+
+        /*
+         * We must xchg() the @node->next value, because if we were to
+         * leave it in, a concurrent unlock()/unqueue() from
+         * @node->next might complete Step-A and think its @prev is
+         * still valid.
+         *
+         * If the concurrent unlock()/unqueue() wins the race, we'll
+         * wait for either @lock to point to us, through its Step-B, or
+         * wait for a new @node->next from its Step-C.
+         */
+        if (node->next) {
+            next = xchg(&node->next, NULL);
+            if (next)
+                break;
+        }
+
+        cpu_relax();
+    }
+
+    return next;
+}
+
+bool osq_lock(struct optimistic_spin_queue *lock)
+{
+    /* sched_getcpu() return current cpu number */
+    int cpu_number = sched_getcpu();
+    if (cpu_number == -1) exit(errno);
+
+    /* each core should have only one thread and one spin_node */
+    struct optimistic_spin_node *node = osq_nodepool_ptr + cpu_number;
+    struct optimistic_spin_node *prev, *next;
+    int curr = encode_cpu(cpu_number);
+    int old;
+    int back_off = 0;
+
+    node->locked = 0;
+    node->next = NULL;
+    node->cpu = curr;
+
+    /*
+     * We need both ACQUIRE (pairs with corresponding RELEASE in
+     * unlock() uncontended, or fastpath) and RELEASE (to publish
+     * the node fields we just initialised) semantics when updating
+     * the lock tail.
+     */
+    old = atomic_xchg(&lock->tail, curr);
+    if (old == OSQ_UNLOCKED_VAL)
+        return true;
+
+    prev = cpu_to_node(old);
+    node->prev = prev;
+
+    /*
+     * osq_lock()            unqueue
+     *
+     * node->prev = prev        osq_wait_next()
+     * WMB                MB
+     * prev->next = node        next->prev = prev // unqueue-C
+     *
+     * Here 'node->prev' and 'next->prev' are the same variable and we need
+     * to ensure these stores happen in-order to avoid corrupting the list.
+     */
+    smp_wmb();
+
+    WRITE_ONCE(prev->next, node);
+
+    /*
+     * Normally @prev is untouchable after the above store; because at that
+     * moment unlock can proceed and wipe the node element from stack.
+     *
+     * However, since our nodes are static per-cpu storage, we're
+     * guaranteed their existence -- this allows us to apply
+     * cmpxchg in an attempt to undo our queueing.
+     */
+
+    while (!READ_ONCE(node->locked)) {
+        /*
+         * If we need to reschedule bail... so we can block.
+         * Use vcpu_is_preempted() to avoid waiting for a preempted
+         * lock holder:
+         */
+        /* TODO: How to emulate rescheduling? */
```

Comment:
I am not entirely sure we can without emulating all of the kernel mutex code which is probably not what we want.

Reply:
We can expose this as extended parameter like previous tbb patch.